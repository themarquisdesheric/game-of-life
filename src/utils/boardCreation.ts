import { Emojis } from '../enums'

const CHANCES_OF_LIFE = 0.3

export const createNewCell = () => ({
  emoji: Emojis.baby,
  ageInGenerations: 1,
})

const createEmptyCell = () => ({
  emoji: Emojis.empty,
  ageInGenerations: 0,
})

export const createDeadCell = () => ({
  emoji: Emojis.dead,
  ageInGenerations: 0,
})

const populateCell = () =>
  (Math.random() <= CHANCES_OF_LIFE)
    ? createNewCell()
    : createEmptyCell()

const createRow = (length: number) => 
  new Array(length)
    .fill(createEmptyCell())
    .map(populateCell)

export const createGameBoard = (length = 5) =>
  new Array(length)
    .fill(createEmptyCell())
    .map(() => createRow(length))
