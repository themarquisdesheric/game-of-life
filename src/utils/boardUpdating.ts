import _isEqual from 'lodash.isequal'
import type { GameBoard, Row, Cell, ComparisonCell, ComparisonGameBoard } from '../global'
import { Emojis } from '../enums'
import { createNewCell, createDeadCell } from './boardCreation'

export const isAlive = (cell: Cell) =>
  cell !== undefined && cell.ageInGenerations

type getNeighborCountFromRowArgs = {
  row: Row,
  targetCellIndex: number,
  rowIncludesTargetCell: boolean
}

const getNeighborCountFromRow = ({
  row,
  targetCellIndex,
  rowIncludesTargetCell = false
}: getNeighborCountFromRowArgs) => {
  let neighborCount = 0;

  for (let i = targetCellIndex - 1; i <= targetCellIndex + 1; i++) {
    if (isAlive(row[i])) {
      if (rowIncludesTargetCell) {
        // and is not target cell
        if (i !== targetCellIndex) {
          neighborCount += 1
        }
      } else {
        neighborCount += 1
      }
    }
  }

  return neighborCount
}

type getNeighborCountArgs = {
  gameBoard: GameBoard,
  y: number,
  x: number
}

const getNeighborCount = ({
  gameBoard,
  y: targetRowIndex,
  x: targetCellIndex,
}: getNeighborCountArgs) => {
  let neighborCount = 0;
  // loop over rows immediately surrounding the target cell 
  for (let i = -1; i <= 1; i++) {
    // if row exists (i.e. target row will not have one above it if it is the top row)
    if (gameBoard[targetRowIndex + i]) {
      const neighborCountFromRow = getNeighborCountFromRow({
        row: gameBoard[targetRowIndex + i],
        targetCellIndex,
        rowIncludesTargetCell: i === 0
      })

      neighborCount += neighborCountFromRow
    }
  }

  return neighborCount
}

const updateSurvivorEmoji = (ageInGenerations: number) => {
  if (ageInGenerations <= 5) return Emojis.baby
  if (ageInGenerations <= 10) return Emojis.man
  if (ageInGenerations <= 15) return Emojis.old
  
  return Emojis.wizard
}

export const updateGameBoard = (prevGameBoard: GameBoard) => {
  const newGameBoard: GameBoard = prevGameBoard.map(row =>
    row.map(cell => ({ ...cell }))
  )

  prevGameBoard.forEach((row, y) => {
    row.forEach((_, x) => {
      const neighborCount = getNeighborCount({
        gameBoard: prevGameBoard,
        y,
        x
      })
      
      if (isAlive(prevGameBoard[y][x])) {
        if (neighborCount < 2) {
          // 💀 any live cell with fewer than two live neighbours dies, as if by underpopulation
          newGameBoard[y][x] = createDeadCell()
        }

        if (neighborCount > 3) {
          // 💀 any live cell with more than three live neighbours dies, as if by overpopulation
          newGameBoard[y][x] = createDeadCell()
        }
        
        if (neighborCount === 2 || neighborCount === 3) {
          // 😀 any live cell with two or three live neighbours lives on to the next generation
          const ageInGenerations = prevGameBoard[y][x].ageInGenerations + 1

          newGameBoard[y][x] = {
            emoji: updateSurvivorEmoji(ageInGenerations),
            ageInGenerations,
          }
        }
        // cell is dead
      } else {
        if (neighborCount === 3) {
          // 🐣 any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction
          newGameBoard[y][x] = createNewCell()
        }

        if (newGameBoard[y][x].emoji === Emojis.dead) {
          newGameBoard[y][x].emoji = Emojis.empty
        }
      }
    })
  })

  return newGameBoard
}

type removeAgeInGenerationsAndFlagCellsArgs = {
  gameBoard: GameBoard,
  checkForNonWizards?: boolean,
  emojiMode: boolean,
}

type removeAgeInGenerationsAndFlagCellsReturnType = {
  gameBoard: ComparisonGameBoard,
  nonWizardFound: boolean,
  liveCellsRemaining: boolean,
}

export const removeAgeInGenerationsAndFlagCells = ({
  gameBoard,
  checkForNonWizards = false,
  emojiMode
}: removeAgeInGenerationsAndFlagCellsArgs
): removeAgeInGenerationsAndFlagCellsReturnType => {
  let nonWizardFound = false
  let liveCellsRemaining = false

  const newGameBoard = gameBoard.map(row => 
    row.reduce(
      (accumulator: ComparisonCell[], cell: Cell) => {
        const isCellAlive = isAlive(cell)

        if (!liveCellsRemaining && isCellAlive) {
          liveCellsRemaining = true
        }

        if (!emojiMode) {
          accumulator.push({
            // retaining emoji structure for subsequent board comparison
            emoji: isCellAlive ? Emojis.baby : Emojis.empty,
          })

          return accumulator
        }
        
        if (checkForNonWizards && isCellAlive) {
          if (cell.emoji !== Emojis.wizard) {
            nonWizardFound = true
          }
        }

        accumulator.push({
          emoji: cell.emoji,
        })

        return accumulator
      }, []
    )
  )

  return {
    gameBoard: newGameBoard,
    nonWizardFound,
    liveCellsRemaining,
  }
}

export const messages = {
  evolutionOver: (generations: number, emojiMode: boolean) => {
    // emoji mode has an extra generation to show skull emoji
    generations = emojiMode ? generations : generations - 1

    return `Conditions were not right for this population to exist. They perished after ${generations} generations.`
  },
  generations: (generations: number) =>
    `${generations} generation${generations > 1 ? 's' : ''}`,
  prosperity: (generations: number) =>
    `It took ${generations} generations for this population to achieve prosperity. Their numbers remain constant, their ranks immortal.`,
  completeWizardhood: 'This population has achieved complete wizardhood. Their numbers remain constant, their ranks immortal.',
  infiniteLoop: 'This population has fallen into an infinite loop.',
}

type isEvolutionOverArgs = {
  newGameBoard: GameBoard,
  previousGameBoardsStore: {
    subscribe: (this: void, run: () => {}, invalidate?: () => {}) => () => {};
    add: (gameBoard: ComparisonGameBoard) => void;
  },
  previousGameBoards: ComparisonGameBoard[],
  generations: number,
  emojiMode: boolean,
}

export const isEvolutionOver = ({ newGameBoard, previousGameBoardsStore, previousGameBoards, generations, emojiMode }: isEvolutionOverArgs) => {
  const {
    gameBoard: newGameBoardNoAgeInGenerations,
    nonWizardFound,
    liveCellsRemaining,
  } = removeAgeInGenerationsAndFlagCells({
    gameBoard: newGameBoard,
    checkForNonWizards: true,
    emojiMode
  })

  if (!liveCellsRemaining) return {
    message: messages.evolutionOver(generations, emojiMode)
  }

  const [nMinusTwoGameBoard, nMinusOneGameBoard] = previousGameBoards
  
  previousGameBoardsStore.add(newGameBoardNoAgeInGenerations)
  
  // if in emoji mode, let evolution continue until remaining living cells have evolved into wizards
  if (emojiMode && nonWizardFound) return false
  
  // evolution has ended when gameboards cease to differ
  const boardsAreEqual = _isEqual(nMinusOneGameBoard, newGameBoardNoAgeInGenerations)
  // one form of infinite loop is recognized when the n - 2 and n gameboards match
  const inInfiniteLoop = _isEqual(nMinusTwoGameBoard, newGameBoardNoAgeInGenerations)

  if (inInfiniteLoop) return {
    message: messages.infiniteLoop,
  }

  if (boardsAreEqual) return {
    message: emojiMode ? messages.completeWizardhood : messages.prosperity(generations)
  }
  
  return false
}
