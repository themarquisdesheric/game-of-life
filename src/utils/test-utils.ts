import type { GameBoard } from '../global'
import { createEmptyCell, createRow } from './boardCreation'

const BOARD_LENGTH = 3

export const createEmptyGameBoard = (boardLength = BOARD_LENGTH): GameBoard =>
  new Array(boardLength)
  .fill(createEmptyCell())
  .map(() => createRow(boardLength, true))
