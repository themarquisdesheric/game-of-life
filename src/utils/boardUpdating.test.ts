import { createNewCell } from './boardCreation'
import { getNeighborCount, updateGameBoard } from './boardUpdating'
import { createEmptyGameBoard } from './test-utils'

describe('getNeighborCount', () => {
  test('empty board should return zero neighbors', () => {
    const gameBoard = createEmptyGameBoard()
    
    const neighborCount = getNeighborCount({
      gameBoard,
      x: 1,
      y: 1,
    })

    expect(neighborCount).toBe(0)
  })

  test('should return one neighbor', () => {
    const gameBoard = createEmptyGameBoard()
    
    gameBoard[0][0] = createNewCell()

    const neighborCount = getNeighborCount({
      gameBoard,
      x: 1,
      y: 1,
    })

    expect(neighborCount).toBe(1)
  })

  test('should return two neighbors', () => {
    const gameBoard = createEmptyGameBoard()
    
    gameBoard[0][0] = createNewCell()
    gameBoard[2][2] = createNewCell()

    const neighborCount = getNeighborCount({
      gameBoard,
      x: 1,
      y: 1,
    })

    expect(neighborCount).toBe(2)
  })
})

describe('updateGameBoard', () => {
  test('identity: i => i', () => {
    /*
      gameBoard: [
        [0, 0, 0],
        [0, 0, 0],
        [0, 0, 0],
      ]
    */
    const gameBoard = createEmptyGameBoard()

    const newGameBoard = updateGameBoard(gameBoard)

    expect(gameBoard).toEqual(newGameBoard)
  })

  test('💀 any live cell with fewer than two live neighbours dies, as if by underpopulation (zero neighbors)', () => {
    /*
      gameBoard: [
        [0, 0, 0],
        [0, 1, 0],
        [0, 0, 0],
      ]
    */
    const gameBoard = createEmptyGameBoard()

    gameBoard[1][1] = createNewCell()

    const newGameBoard = updateGameBoard(gameBoard)

    expect(newGameBoard[1][1].ageInGenerations).toBe(0)
    expect(newGameBoard).toMatchSnapshot()
  })

  test('💀 any live cell with fewer than two live neighbours dies, as if by underpopulation (one neighbor)', () => {
    /*
      gameBoard: [
        [1, 0, 0],
        [0, 1, 0],
        [0, 0, 0],
      ]
    */
    const gameBoard = createEmptyGameBoard()

    gameBoard[0][0] = createNewCell()
    gameBoard[1][1] = createNewCell()

    const newGameBoard = updateGameBoard(gameBoard)

    expect(newGameBoard[1][1].ageInGenerations).toBe(0)
    expect(newGameBoard).toMatchSnapshot()
  })

  test('💀 any live cell with more than three live neighbours dies, as if by overpopulation', () => {
    /*
      gameBoard: [
        [1, 1, 1],
        [1, 1, 0],
        [0, 0, 0],
      ]
    */
    const gameBoard = createEmptyGameBoard()

    gameBoard[0][0] = createNewCell()
    gameBoard[0][1] = createNewCell()
    gameBoard[0][2] = createNewCell()
    gameBoard[1][0] = createNewCell()
    gameBoard[1][1] = createNewCell()
 
    const newGameBoard = updateGameBoard(gameBoard)

    expect(newGameBoard[1][1].ageInGenerations).toBe(0)
    expect(newGameBoard).toMatchSnapshot()
  })

  test('😀 any live cell with two or three live neighbours lives on to the next generation (two neighbors)', () => {
    /*
      gameBoard: [
        [1, 1, 0],
        [0, 1, 0],
        [0, 0, 0],
      ]
    */
    const gameBoard = createEmptyGameBoard()

    gameBoard[0][0] = createNewCell()
    gameBoard[0][1] = createNewCell()
    gameBoard[1][1] = createNewCell()

    const newGameBoardWithTwoNeighbors = updateGameBoard(gameBoard)

    expect(newGameBoardWithTwoNeighbors[1][1].ageInGenerations).toBe(2)
    expect(newGameBoardWithTwoNeighbors).toMatchSnapshot()

    /*
      gameBoard: [
        [1, 1, 1],
        [0, 1, 0],
        [0, 0, 0],
      ]
    */
    gameBoard[0][2] = createNewCell()

    const newGameBoardWithThreeNeighbors = updateGameBoard(gameBoard)

    expect(newGameBoardWithThreeNeighbors[1][1].ageInGenerations).toBe(2)
    expect(newGameBoardWithThreeNeighbors).toMatchSnapshot()
  })

  test('🐣 any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction', () => {
    /*
      gameBoard: [
        [0, 1, 0],
        [1, 1, 0],
        [0, 0, 0],
      ]
    */
    const gameBoard = createEmptyGameBoard()

    gameBoard[0][1] = createNewCell()
    gameBoard[1][0] = createNewCell()
    gameBoard[1][1] = createNewCell()

    const newGameBoard = updateGameBoard(gameBoard)
    
    expect(newGameBoard[0][0].ageInGenerations).toBe(1)
    expect(newGameBoard).toMatchSnapshot()
  })
})

