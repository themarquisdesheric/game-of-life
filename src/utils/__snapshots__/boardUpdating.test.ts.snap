// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateGameBoard 🐣 any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction 1`] = `
Array [
  Array [
    Object {
      "ageInGenerations": 1,
      "emoji": "👶🏻",
    },
    Object {
      "ageInGenerations": 2,
      "emoji": "👶🏻",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
  ],
  Array [
    Object {
      "ageInGenerations": 2,
      "emoji": "👶🏻",
    },
    Object {
      "ageInGenerations": 2,
      "emoji": "👶🏻",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
  ],
  Array [
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
  ],
]
`;

exports[`updateGameBoard 💀 any live cell with fewer than two live neighbours dies, as if by underpopulation (one neighbor) 1`] = `
Array [
  Array [
    Object {
      "ageInGenerations": 0,
      "emoji": "💀",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
  ],
  Array [
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "💀",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
  ],
  Array [
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
  ],
]
`;

exports[`updateGameBoard 💀 any live cell with fewer than two live neighbours dies, as if by underpopulation (zero neighbors) 1`] = `
Array [
  Array [
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
  ],
  Array [
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "💀",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
  ],
  Array [
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
  ],
]
`;

exports[`updateGameBoard 💀 any live cell with more than three live neighbours dies, as if by overpopulation 1`] = `
Array [
  Array [
    Object {
      "ageInGenerations": 2,
      "emoji": "👶🏻",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "💀",
    },
    Object {
      "ageInGenerations": 2,
      "emoji": "👶🏻",
    },
  ],
  Array [
    Object {
      "ageInGenerations": 2,
      "emoji": "👶🏻",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "💀",
    },
    Object {
      "ageInGenerations": 1,
      "emoji": "👶🏻",
    },
  ],
  Array [
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
  ],
]
`;

exports[`updateGameBoard 😀 any live cell with two or three live neighbours lives on to the next generation (two neighbors) 1`] = `
Array [
  Array [
    Object {
      "ageInGenerations": 2,
      "emoji": "👶🏻",
    },
    Object {
      "ageInGenerations": 2,
      "emoji": "👶🏻",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
  ],
  Array [
    Object {
      "ageInGenerations": 1,
      "emoji": "👶🏻",
    },
    Object {
      "ageInGenerations": 2,
      "emoji": "👶🏻",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
  ],
  Array [
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
  ],
]
`;

exports[`updateGameBoard 😀 any live cell with two or three live neighbours lives on to the next generation (two neighbors) 2`] = `
Array [
  Array [
    Object {
      "ageInGenerations": 2,
      "emoji": "👶🏻",
    },
    Object {
      "ageInGenerations": 2,
      "emoji": "👶🏻",
    },
    Object {
      "ageInGenerations": 2,
      "emoji": "👶🏻",
    },
  ],
  Array [
    Object {
      "ageInGenerations": 1,
      "emoji": "👶🏻",
    },
    Object {
      "ageInGenerations": 2,
      "emoji": "👶🏻",
    },
    Object {
      "ageInGenerations": 1,
      "emoji": "👶🏻",
    },
  ],
  Array [
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
    Object {
      "ageInGenerations": 0,
      "emoji": "",
    },
  ],
]
`;
